tests/suites/test_suite_platform.function

```diff
@@ -32,91 +32,106 @@ void sleep_ms(int milliseconds)
     usleep(milliseconds * 1000);
 #endif
 }
-#endif
 
+#endif
 /* END_HEADER */
 
 /* BEGIN_DEPENDENCIES */
 
 /* END_DEPENDENCIES */
 
 /* BEGIN_CASE depends_on:MBEDTLS_HAVE_TIME */
-void time_get_milliseconds()
-{
-    mbedtls_ms_time_t current = mbedtls_ms_time();
-    (void) current;
-    /* This goto is added to avoid warnings from the generated code. */
-    goto exit;
-}
-/* END_CASE */
-
-/* BEGIN_CASE depends_on:MBEDTLS_HAVE_TIME */
-void time_get_seconds()
-{
-    mbedtls_time_t current = mbedtls_time(NULL);
-    (void) current;
-    /* This goto is added to avoid warnings from the generated code. */
-    goto exit;
-}
-/* END_CASE */
 
-/* BEGIN_CASE depends_on:MBEDTLS_HAVE_TIME */
+/* `time_delay_milliseconds()` and `time_delay_seconds()` might be fail due to
+ * `signal interrupt`, `discontinues jumps` and `incremental adjustment`. The
+ * test functions run 3 iteration to ignore the failures caused by those
+ * conditions.
+ * - `signal interrupt`  : That is introduced by `sleep_ms()`. `nanosleep()` and
+ *                         `usleep()` can be interrupted by signal.
+ * - `discontinues jumps`: `CLOCK_REALTIME` can be manually changed. When it
+ *                         happens during test, the test will be failed on
+ *                         Linux.
+ * - `incremental adjustment`: When `incremental adjustments` are performed
+ *                         during test, the end time might be not in tolerance
+ *                         windows. If `incremental adjustments` end in 2nd
+ *                         iteration, it might fail also, we try 3rd iteration.
+ *
+ * `time_delay_milliseconds()` and `time_delay_seconds()` will re-try 3 times,
+ * so we design a +/- 1 second expect window.
+ */
 void time_delay_milliseconds(int delay_ms)
 {
-    mbedtls_ms_time_t current = mbedtls_ms_time();
-    mbedtls_ms_time_t elapsed_ms;
-
-    /*
-     * WARNING: DO NOT ENABLE THIS TEST. We keep the code here to document the
-     *          reason.
-     *
-     *          Windows CI reports random test fail on platform-suite. It might
-     *          be caused by this case.
-     */
-    sleep_ms(delay_ms);
+    mbedtls_ms_time_t current;
+    mbedtls_ms_time_t expected_end;
+    mbedtls_ms_time_t diff_ms;
 
-    elapsed_ms = mbedtls_ms_time() - current;
-    TEST_ASSERT(elapsed_ms >= delay_ms && elapsed_ms < 4000 + delay_ms);
-    /* This goto is added to avoid warnings from the generated code. */
-    goto exit;
+    current = mbedtls_ms_time();
+    expected_end = current + delay_ms;
+    sleep_ms(delay_ms);
+    diff_ms = mbedtls_ms_time() - expected_end;
+    if (diff_ms < 1000 && -1000 < diff_ms) {
+        goto exit;
+    }
+
+    /* When `signal interrupt`, `discontinues jumps` or `incremental adjustment`
+     * happens in previous iteration, try again.*/
```

Comment:
We used to have tests of timing functions like this one. The reason they failed was that the CI machine was overloaded and the process was sometimes not scheduled for over 1 second, so `time(); sleep(); time()` reported a delay that was much longer than expected. This new code has the same defect.

Retrying doesn't really help, because if the system load was high during the first iteration, it's very likely to still be high during the second iteration.

Given that this reintroduces a problem that we fixes, this is a blocker.

Reply:
I do not expect it can resolve heavy loading reason.